events/importer/espoo.py

```diff
@@ -1,800 +1,556 @@
+    - instance_map: a map of origin object ids to target (local) object ids
+    and return a new dict of fields to be used for the update of the target object
+    """,
+]
+
+
+class EspooImporterError(Exception):
+    pass
+
+
+def parse_jsonld_id(data: dict) -> str:
+    if (ld_reference := data.get("@id")) and (origin_id := ld_reference.split("/")[-2]):
+        return origin_id
+    raise EspooImporterError("Could not parse jsonld id")
+
+
+def parse_jsonld_ids(data: list) -> list[str]:
+    return [parse_jsonld_id(d) for d in data]
+
+
+def _get_data(url: str, params: Optional[dict[str, str]] = None) -> dict:
+    for _ in range(settings.ESPOO_MAX_RETRIES):
+        response = requests.get(url, timeout=settings.ESPOO_TIMEOUT, params=params)
+        if response.status_code == 200:
+            return response.json()
+        elif response.status_code < 500:
+            raise EspooImporterError(
+                f"ESPOO API returned status code {response.status_code}"
+            )
+        time.sleep(settings.ESPOO_WAIT_BETWEEN)
+
+    raise EspooImporterError("Exceeded ESPOO_MAX_RETRIES")
+
+
+def _list_data(url: str, params: Optional[dict[str, str]] = None) -> list[dict]:
+    results = []
+    for _ in range(settings.ESPOO_MAX_PAGES):
+        response_data = _get_data(url, params)
+        results += response_data["data"]
+        url = response_data["meta"]["next"]
+        if url is None:
+            return results
+        time.sleep(settings.ESPOO_WAIT_BETWEEN)
+
+    raise EspooImporterError("Exceeded ESPOO_MAX_PAGES")
+
+
+def _get_origin_objs(detail_url: str, origin_obj_ids: list) -> list:
+    return [_get_data(urljoin(detail_url, f"./{obj_id}/")) for obj_id in origin_obj_ids]
+
+
+def _build_pre_map(mapping: dict[str, str], jsonld=False) -> PreMapper:
+    def pre_map(field_name, field_data: Union[str, dict], data: dict) -> dict[str, Any]:
+        if field_data is None:
+            return {**data}
+        if jsonld:
+            return {**data, field_name: mapping[parse_jsonld_id(field_data)]}
+        return {**data, field_name: mapping[field_data]}
+
+    return pre_map
+
+
+def _build_pre_map_to_id(mapping: dict[str, str], jsonld=False) -> PreMapper:
+    """
+    Pre mapper for ForeignKey fields
+    """
+    pre_mapper = _build_pre_map(mapping, jsonld)
+
+    def pre_map_to_id(
+        field_name: str, field_data: Union[str, dict], data: dict
+    ) -> dict[str, Any]:
+        return pre_mapper(f"{field_name}_id", field_data, data)
+
+    return pre_map_to_id
+
+
+def _pre_map_translation(
+    field_name: str, field_data: Optional[dict], data: dict
+) -> dict:
+    """
+    Translation field mapper
+    """
+    if field_data is None:
+        return data
+
+    # LANGUAGES is a tuple containing (code, label) pairs
+    # [("fi", "Finnish"), ...]
+    # zip it into a list containing a tuple of codes and a tuple of labels
+    # [("fi", ...), ("Finnish", ...)]
+    language_codes = list(zip(*settings.LANGUAGES))[0]
```

Comment:
```suggestion
    language_codes = [lang[0] for lang in settings.LANGUAGES]
```
I think this would be more readable and doesn't necessarily require a comment.

Reply:
![caseclosed](https://github.com/City-of-Helsinki/linkedevents/assets/437576/7967366c-8e37-4603-bee7-bb3a1baa57e5)